src/core/thread/mle.cpp

```diff
@@ -2033,49 +2035,125 @@ otError Mle::SendDataRequest(const Ip6::Address &aDestination,
         message->Free();
     }
 
-    if (!IsRxOnWhenIdle())
+    if ((mRole == OT_DEVICE_ROLE_CHILD) && !IsRxOnWhenIdle())
     {
-        mChildUpdateRequestTimer.Start(kMaxChildUpdateResponseTimeout);
+        mDataRequestState = kDataRequestActive;
+
+        if (mChildUpdateRequestState == kChildUpdateRequestNone)
+        {
+            ScheduleMessageTransmissionTimer();
+        }
     }
 
     return error;
 }
 
-void Mle::HandleChildUpdateRequestTimer(Timer &aTimer)
+void Mle::ScheduleMessageTransmissionTimer(void)
 {
-    aTimer.GetOwner<Mle>().HandleChildUpdateRequestTimer();
+    uint32_t interval = 0;
+
+    switch (mChildUpdateRequestState)
+    {
+    case kChildUpdateRequestNone:
+        break;
+
+    case kChildUpdateRequestPending:
+        ExitNow(interval = kChildUpdateRequestPendingDelay);
+
+    case kChildUpdateRequestActive:
+        ExitNow(interval = kUnicastRetransmissionDelay);
+    }
+
+    switch (mDataRequestState)
+    {
+    case kDataRequestNone:
+        break;
+
+    case kDataRequestActive:
+        ExitNow(interval = kUnicastRetransmissionDelay);
+    }
+
+    if ((mRole == OT_DEVICE_ROLE_CHILD) && IsRxOnWhenIdle())
+    {
+        interval = TimerMilli::SecToMsec(mTimeout) -
+                   static_cast<uint32_t>(kUnicastRetransmissionDelay) * kMaxChildKeepAliveAttempts;
+    }
+
+exit:
+    if (interval != 0)
+    {
+        mMessageTransmissionTimer.Start(interval);
+    }
+    else
+    {
+        mMessageTransmissionTimer.Stop();
+    }
 }
 
-void Mle::HandleChildUpdateRequestTimer(void)
+void Mle::HandleMessageTransmissionTimer(Timer &aTimer)
 {
-    // Here intentionly delay another kChildUpdateRequestPending cycle to ensure
-    // only send a Child Update Request after we know there are no more pending changes
+    aTimer.GetOwner<Mle>().HandleMessageTransmissionTimer();
+}
+
+void Mle::HandleMessageTransmissionTimer(void)
+{
+    // The `mMessageTransmissionTimer` is used for:
+    //
+    //  - Delaying OT_CHANGED notification triggered "Child Update Request" transmission (to allow aggregation),
+    //  - Retransmission of "Child Update Request",
+    //  - Retransmission of "Data Request" on a child,
+    //  - Sending periodic keep-alive "Child Update Request" messages on a non-sleepy (rx-on) child.
+
     if (mChildUpdateRequestState == kChildUpdateRequestPending)
     {
         if (GetNotifier().IsPending())
         {
-            mChildUpdateRequestTimer.Start(kChildUpdateRequestPendingDelay);
+            // Here intentionally delay another kChildUpdateRequestPendingDelay
+            // cycle to ensure we only send a Child Update Request after we
+            // know there are no more pending changes.
+            ScheduleMessageTransmissionTimer();
```

Comment:
Move this logic into the switch statement below?

Reply:
It break the fall through logic from `case kChildUpdateRequestNone` into `case kChildUpdateRequestPending`. I think we would need an `if` somewhere. Let me check this a bit more.

Reply:
I was thinking that the if-statement on L2109 would move as well.

Reply:
Thanks. I think figured out a good way with no extra `if` checks. Moved the code for transmission of "Child Update" after the `switch`. This way, we either jump to `exit` or simply break from different `case`s to send the "Child Update". It's cleaner.  Will push it soon.